### PR TITLE
IN-16 Dashboard parts

### PR DESCRIPTION
**Pull request by glachezar**

### CODE REVIEW

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserLatestInvestmentsQuery.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+        operationResult.RelatedObject = result;
+
+        return operationResult;
+    }
+}
+
+public record GetUserLatestInvestmentsQueryModel
+{
+    public int Id { get; set; }
```
Comment (MihaelPavlov) [inline investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserLatestInvestmentsQuery.cs:48]: Here Let's Leave it as AssetId,Name, Date and Amount

Comment (MihaelPavlov) [reply investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserLatestInvestmentsQuery.cs:48]: Return the amount in correct format. Currently is in string, return it as decimal. 
decima.,Parse(x.PurchasePrice)

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserLatestInvestmentsQuery.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
+﻿namespace TRINV.Application.Queries.Dashboard;
+
+using Domain.Entities;
+using Interfaces;
+using MediatR;
+using Shared.Business.Utilities;
+
+public record GetUserLatestInvestmentsQuery : IRequest<OperationResult<IList<GetUserLatestInvestmentsQueryModel>>>;
```
Comment (MihaelPavlov) [inline investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserLatestInvestmentsQuery.cs:8]: You need to accepect Investment Type and use it inside currentUserInvestments, Currently you are returning all investment for this user, We need it based on the investment Type

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
+﻿namespace TRINV.Application.Queries.Dashboard;
+
+using DigitalCurrency;
+using Domain.Entities;
+using Domain.Enums;
+using Interfaces;
+using MediatR;
+using Shared.Business.Utilities;
+
+public record GetInvestmentPerformanceQuery : IRequest<OperationResult<List<InvestmentPerformanceQueryModel>>>;
```
Comment (MihaelPavlov) [inline investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs:10]: The same here you need to accept Investment Type and use it in currentUserInvestments

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            .ToList();
+
+        operationResult.RelatedObject = currentUserInvestmentsPerformanceList;
+
+        return operationResult;
+    }
+}
+
+public record InvestmentPerformanceQueryModel
+{
```
Comment (MihaelPavlov) [inline investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs:84]: Let's leave the properties as this : 
AssetId, Name, TotalInitialInvestment, TotalCurrentInvestment,Rate.

Leave  TotalInitialInvestment, TotalCurrentInvestment and Rate. as decimal

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        _investmentRepository = investmentRepository;
+        _userContext = userContext;
+        _mediator = mediator;
+    }
+
+    public async Task<OperationResult<List<InvestmentPerformanceQueryModel>>> Handle(GetInvestmentPerformanceQuery request, CancellationToken cancellationToken)
+    {
+        var operationResult = new OperationResult<List<InvestmentPerformanceQueryModel>>();
+
+        var currentUserInvestments = await _investmentRepository
```
Comment (MihaelPavlov) [inline investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs:30]: Did you test it is it work correctly ? 
If not Change the _userContext.userId to 1 and add investments with userId 1.

File: investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/DashboardController.cs [hunk 6]
```diff
@@ -33,4 +33,20 @@ public async Task<IActionResult> GetInvesmentsInPercents([FromQuery] GetDashboar
         var result = await _mediator.Send(query, cancellationToken);
         return Ok(result);
     }
+
+    [HttpGet("get-investments-performance")]
```
Comment (MihaelPavlov) [inline investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/DashboardController.cs:37]: i think get-investment-performance-list sound better

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserLatestInvestmentsQuery.cs (hunk 1 above)
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserLatestInvestmentsQuery.cs:48]: Done.

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserLatestInvestmentsQuery.cs (hunk 2 above)
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserLatestInvestmentsQuery.cs:8]: Done

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs (hunk 3 above)
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs:10]: Done.

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs (hunk 4 above)
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs:84]: Done

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs (hunk 5 above)
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs:30]: Yes, it is tested!

File: investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/DashboardController.cs (hunk 6 above)
Comment (glachezar) [reply investment-app/TRINV.Api/TRINV.InvestTrackApplication/Controllers/DashboardController.cs:37]: Done

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            };
+
+            currentUserInvestmentsPerformanceList.Add(result);
+        }
+
+        currentUserInvestmentsPerformanceList = currentUserInvestmentsPerformanceList
+            .OrderByDescending(x => x.Rate)
+            .ToList();
+
+        operationResult.RelatedObject = currentUserInvestmentsPerformanceList;
```
Comment (MihaelPavlov) [inline investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs:92]: You can remove  line 88 and do it like this operationResult.RelatedObject = currentUserInvestmentsPerformanceList.OrderBy(x=>X..rate);
 
I'm just mentioning it., Don't spend time i will fix it when i am adjusting the fe

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                AverageValuePerUnitInFiat = x.Sum(i => i.PurchasePrice) / x.Sum(q => q.Quantity),
+            });
+
+        Dictionary<string, decimal> currentPrices = new Dictionary<string, decimal>();
+
+        if(request.InvestmentType == InvestmentType.Cryptocurrency)
+        {
+            var currentCryptoPrices = await _mediator.Send(new GetDigitalCurrencyListQuery(), cancellationToken);
+
+            foreach (var item in groupedInvestments)
```
Comment (MihaelPavlov) [inline investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs:57]: Here we can remove the foreach and do it something like ''

 currentPrices = currentCryptoPrices.RelatedObject
        .Where(s => groupedInvestments.Any(gi => gi.Symbol == s.Symbol))
        .ToDictionary(s => s.Symbol, p => decimal.Parse(p.PriceUsd));

File: investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                    .Select(p =>p.PriceUsd)
+                    .FirstOrDefault());
+
+                currentPrices.Add(item.Symbol, currentItemPrice);
+            }
+        }
+
+        var currentUserInvestmentsPerformanceList = new List<InvestmentPerformanceQueryModel>();
+
+        foreach (var investment in groupedInvestments)
```
Comment (MihaelPavlov) [inline investment-app/TRINV.Api/TRINV.Application/Queries/Dashboard/GetUserInvestmentsPerformanceListQuery.cs:70]: And here again we can remove the foreach and do the calculation for example in select 
Something like that -> 

var currentUserInvestmentsPerformanceList = groupedInvestments
    .Select(investment =>
    {
        var currentInvestmentPrice = currentPrices.GetValueOrDefault(investment.Symbol, 0m);

        var returnRate = ((currentInvestmentPrice - investment.AverageValuePerUnitInFiat) / investment.AverageValuePerUnitInFiat) * 100;

        return new InvestmentPerformanceQueryModel
        {
            AssetId = investment.Symbol,
            Name = investment.Name,
            TotalInitialInvestment = Math.Round(investment.TotalValueInFiat, 2),
            TotalCurrentInvestment = Math.Round(currentInvestmentPrice, 2),
            Rate = Math.Round(returnRate, 2)
        };
    })
    .OrderByDescending(x => x.Rate)
    .ToList();

**Review by MihaelPavlov — APPROVED** [review]

**Pull request merged**